components/raftstore/src/store/async_io.rs

```diff
@@ -0,0 +1,949 @@
+// Copyright 2021 TiKV Project Authors. Licensed under Apache-2.0.
+
+// TODO: move it
+#![allow(dead_code)]
+use std::marker::PhantomData;
+use std::sync::mpsc::{channel, Receiver, Sender, TryRecvError};
+use std::thread::{self, JoinHandle};
+use std::time::Instant;
+
+use crate::store::config::Config;
+use crate::store::fsm::RaftRouter;
+use crate::store::local_metrics::RaftSendMessageMetrics;
+use crate::store::metrics::*;
+use crate::store::transport::Transport;
+use crate::store::{PeerMsg, SignificantMsg};
+use crate::Result;
+
+use collections::{HashMap, HashSet};
+use engine_traits::{
+    KvEngine, PerfContext, PerfContextKind, RaftEngine, RaftLogBatch, WriteBatch, WriteOptions,
+};
+use error_code::ErrorCodeExt;
+use fail::fail_point;
+use kvproto::raft_serverpb::{RaftLocalState, RaftMessage};
+use raft::eraftpb::Entry;
+use tikv_util::time::duration_to_sec;
+use tikv_util::{box_err, debug, thd_name, warn};
+
+const KV_WB_SHRINK_SIZE: usize = 1024 * 1024;
+const RAFT_WB_SHRINK_SIZE: usize = 10 * 1024 * 1024;
+
+/// Notify the event to the specified region.
+pub trait Notifier: Clone + Send + 'static {
+    fn notify_persisted(&self, region_id: u64, peer_id: u64, ready_number: u64, now: Instant);
+    fn notify_unreachable(&self, region_id: u64, to_peer_id: u64);
+}
+
+#[derive(Clone)]
+pub struct RegionNotifier<EK, ER>
+where
+    EK: KvEngine,
+    ER: RaftEngine,
+{
+    router: RaftRouter<EK, ER>,
+}
+
+impl<EK, ER> RegionNotifier<EK, ER>
+where
+    EK: KvEngine,
+    ER: RaftEngine,
+{
+    pub fn new(router: RaftRouter<EK, ER>) -> Self {
+        Self { router }
+    }
+}
+
+impl<EK, ER> Notifier for RegionNotifier<EK, ER>
```

Comment:
Why not implement `Notifier` for `RaftRouter` directly?

Reply:
Makes sense🤣.